mobile_verifier/src/subsciber_location.rs

```diff
@@ -0,0 +1,179 @@
+            .try_fold(transaction, |mut transaction, report| async move {
+                if self.verify_known_carrier_key(&report.report.pubkey) {
+                    self.save(
+                        report.report.subscriber_id.clone(),
+                        report.received_timestamp,
+                        &mut transaction,
+                    )
+                    .await?;
+                    metrics::increment_counter!(
+                        "oracles_mobile_verifier_valid_subscriber_location_report"
+                    );
+                } else {
+                    metrics::increment_counter!(
+                        "oracles_mobile_verifier_invalid_subscriber_location_report"
+                    );
+                }
+                Ok(transaction)
+            })
+            .await?
+            .commit()
+            .await?;
+        Ok(())
+    }
+
+    pub async fn save(
+        &self,
+        subscriber_id: Vec<u8>,
+        timestamp: DateTime<Utc>,
+        db: &mut Transaction<'_, Postgres>,
+    ) -> Result<(), sqlx::Error> {
+        sqlx::query(
+            r#"
+                INSERT INTO subscriber_loc (subscriber_id, date_bucket, hour_bucket, reward_timestamp)
+                VALUES ($1, DATE_TRUNC('hour', $2), DATE_PART('hour', $2), $2)
+                ON CONFLICT DO NOTHING;
+                "#,
+        )
+        .bind(subscriber_id)
+        .bind(timestamp)
+        .execute(&mut *db)
+        .await?;
+        Ok(())
+    }
+
+    //TODO: reinstate check when keys are available
+    fn verify_known_carrier_key(&self, _public_key: &PublicKeyBinary) -> bool {
+        // self.carrier_keys.contains(public_key)
+        true
+    }
+
+    // TODO: update when actual API is available
+    async fn refresh_carrier_keys(&mut self) -> anyhow::Result<()> {
+        // match self
+        // .config_client
+        // .clone()
+        // .resolve_carrier_keys()
+        // .await
+        // {
+        //     Ok(res) => {
+        //         self.carrier_keys = res;
+        //     }
+        //     Err(err) => (),
+        // }
+        Ok(())
+    }
+}
+
+#[derive(sqlx::FromRow)]
+pub struct SubscriberLocationShare {
+    pub subscriber_id: Vec<u8>,
+    pub hour_bucket: i32,
+}
+
+pub async fn aggregate_location_shares(
+    db: impl sqlx::PgExecutor<'_> + Copy,
+    reward_period: &Range<DateTime<Utc>>,
+) -> Result<LocationSharingMap, sqlx::Error> {
+    let mut rows = sqlx::query_as::<_, SubscriberLocationShare>(
+        "select subscriber_id, hour_bucket::int from subscriber_loc where reward_timestamp >= $1 and reward_timestamp < $2",
+    )
+    .bind(reward_period.start)
+    .bind(reward_period.end)
+    .fetch(db);
+    let mut location_shares = LocationSharingMap::new();
+    while let Some(share) = rows.try_next().await? {
+        location_shares
+            .entry(share.subscriber_id)
+            .or_insert_with(Vec::new)
+            .push(share.hour_bucket)
```

Comment:
are these guaranteed to be in order based on ingest? is there any chance we'd need to order this vec right before you call `check_subscriber_location_shares_rewardable`?

Reply:
sorting ensured here: afb0bf72752c1d9c1e2ce2c096831dd2f07e652d